File: frontend/src/features/contacts/api/contacts.ts
import { apiFetch } from '../../../lib/api'

export type ContactType = 'customer' | 'supplier' | 'both'
export type DocumentType = 'ruc' | 'ci' | 'passport' | 'none'

export interface ContactOut {
  id: string
  contact_type: ContactType
  document_type: DocumentType
  document_number: string | null
  business_name: string
  trade_name: string | null
  phone: string | null
  email: string | null
  address: string | null
  notes: string | null
  is_active: boolean
  created_at: string
  updated_at: string
  deleted_at: string | null
  created_by_user_id: string | null
  updated_by_user_id: string | null
}

export interface ContactListOut {
  items: ContactOut[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface ContactCreate {
  id: string
  contact_type: ContactType
  document_type?: DocumentType
  document_number?: string | null
  business_name: string
  trade_name?: string | null
  phone?: string | null
  email?: string | null
  address?: string | null
  notes?: string | null
  is_active?: boolean
}

export interface ContactUpdate {
  contact_type?: ContactType
  document_type?: DocumentType
  document_number?: string | null
  business_name?: string
  trade_name?: string | null
  phone?: string | null
  email?: string | null
  address?: string | null
  notes?: string | null
  is_active?: boolean
}

export interface ContactListParams {
  contact_type?: ContactType
  search?: string
  page?: number
  page_size?: number
}

export function fetchContacts(params: ContactListParams = {}): Promise<ContactListOut> {
  const qs = new URLSearchParams()
  if (params.contact_type) qs.set('contact_type', params.contact_type)
  if (params.search) qs.set('search', params.search)
  if (params.page != null) qs.set('page', String(params.page))
  if (params.page_size != null) qs.set('page_size', String(params.page_size))
  const query = qs.toString()
  return apiFetch<ContactListOut>(`/contacts${query ? `?${query}` : ''}`)
}

export function fetchContact(id: string): Promise<ContactOut> {
  return apiFetch<ContactOut>(`/contacts/${id}`)
}

export function createContact(data: ContactCreate): Promise<ContactOut> {
  return apiFetch<ContactOut>('/contacts', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export function updateContact(id: string, data: ContactUpdate): Promise<ContactOut> {
  return apiFetch<ContactOut>(`/contacts/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  })
}

export function deleteContact(id: string): Promise<void> {
  return apiFetch<void>(`/contacts/${id}`, { method: 'DELETE' })
}

File: backend/app/schemas/contacts.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.enums import ContactType, DocumentType


class ContactCreate(BaseModel):
    id: UUID
    contact_type: ContactType
    document_type: DocumentType = DocumentType.NONE
    document_number: str | None = None
    business_name: str
    trade_name: str | None = None
    phone: str | None = None
    email: str | None = None
    address: str | None = None
    notes: str | None = None
    is_active: bool = True


class ContactUpdate(BaseModel):
    contact_type: ContactType | None = None
    document_type: DocumentType | None = None
    document_number: str | None = None
    business_name: str | None = None
    trade_name: str | None = None
    phone: str | None = None
    email: str | None = None
    address: str | None = None
    notes: str | None = None
    is_active: bool | None = None


class ContactOut(BaseModel):
    id: UUID
    contact_type: ContactType
    document_type: DocumentType
    document_number: str | None
    business_name: str
    trade_name: str | None
    phone: str | None
    email: str | None
    address: str | None
    notes: str | None
    is_active: bool
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None
    created_by_user_id: UUID | None
    updated_by_user_id: UUID | None


class ContactListOut(BaseModel):
    items: list[ContactOut]
    total: int
    page: int
    page_size: int
    total_pages: int
